[Apps] Actions on models

snoe opened the pull request:
[product doc](https://www.notion.so/metabase/Make-models-contain-actions-579624779cb74b23bc58be7478b37d26)
[dev doc](https://www.notion.so/metabase/Actions-in-models-dev-doc-5f038c453fbc43f192f2ba2bd4979069)

The goal with this PR is twofold:
1. Attach actions to models
2. Do so in a low touch way

This path, allowed us to avoid changing `action` or `report_card` tables or models at all. The dev doc has a phase 2 path, that would allow the removal of this `model_action` table but will take too long to get a prototype out for this new concept. Although this table schema allows actions to be placed on multiple actions, it should not be allowed. As well, actions should always be attached to a `model_action`.

In some places, intermediate work is done to smear the line between `action` and `model_action` and present the combined concept as an `action`.  This is evident on `dashcard.action` and `GET /api/action`. This provides a cleaner path towards phase 2.

snoe committed:
[Apps] Model Actions

WIP Adding migration and endpoints for model_action.

snoe committed:
WIP model action execution

Changing the execution endpoints again, still working on implicit action
execution tests. Need to also test the parameters fetch.

snoe committed:
Merge remote-tracking branch 'origin/master' into data-apps-model-actions

snoe committed:
Re-add execution route for dashcard action_id until Front End catches up

snoe committed:
Add name to model_action GET

snoe committed:
Hydrate model-action on dashcards

snoe committed:
Rename and combine hydration of model_action to action

snoe committed:
Implicit action execution support. Use slug columns as parameter ids

snoe committed:
Go through model-action when using GET /action

snoe committed:
Merge remote-tracking branch 'origin/master' into data-apps-model-actions

snoe committed:
FE integration fixes

snoe committed:
Move action execution to use new parameter shape

Now, parameters should be a map of parameter-id to value
Parameter-id should be mapped on the FE. So the action parameter-id not the unmapped
dashboard parameter-id.

snoe committed:
Bring ModelAction into copy code

snoe committed:
Generate target of implicit actions to match http actions

snoe committed:
Update action test to place action on model

snoe committed:
Add unique constraint to model_action for (card_id, slug)

snoe committed:
Add missing require

snoe committed:
model_action.slug needs to be varchar for mysql constraint

snoe committed:
Fix test from unique constraint

snoe committed:
Consistent ordering for tests

metamben commented on test/metabase/api/model_action_test.clj:
```diff
@@ -0,0 +1,43 @@
+                                                :slug "custom"})]
+            (is (partial=
+                  {:slug "custom"}
+                  response))))))))
```
Might want to add tests for the `PUT` endpoint too.

metamben reviewed (commented):
I've left a bunch of questions.

snoe commented on src/metabase/actions/execution.clj:
```diff
@@ -77,50 +44,104 @@
-                             (pr-str (set (keys destination-parameters-by-target))))
+        destination-parameters-by-id (m/index-by :id (:parameters action))]
+    (doseq [[parameter-id _value] request-parameters]
+      (when-not (contains? destination-parameters-by-id parameter-id)
+        (throw (ex-info (tru "No destination parameter found for id {0}. Found: {1}"
+                             (pr-str parameter-id)
+                             (pr-str (set (keys destination-parameters-by-id))))
                         {:status-code 400
                          :type qp.error-type/invalid-parameter
-                         :parameters parameters
+                         :parameters request-parameters
                          :destination-parameters (:parameters action)}))))
```
I put this in to help with "did you forget to map a parameter", now that the dashcard parameter_mapping function moved to the frontend. 

If the client is sending extra parameters that can be ignored, that probably means the action changed from under them. I still think it's safer to error in that case.

Is there a situation where being flexible with input here is wanted?

metamben replied:
I don't have any particular example. I'm just nor sure why we should error if an action changed but is still compatible. I think this a product question after all.

snoe replied:
Ok, I'm going to leave this and we can revisit/remove once it merges.

snoe commented on src/metabase/models/action.clj:
```diff
@@ -152,12 +167,76 @@
-    dashcards))
+  (let [model-slug-by-dashcard-id (->> dashcards
+                                       (keep (fn [dashcard]
+                                               (when-let [slug (get-in dashcard [:visualization_settings :action_slug])]
```
Yeah, this one of the tradeoffs we purposely made for this iteration of model actions.

metamben replied:
Fair enough. Do we track these compromises somewhere?

metamben commented on src/metabase/actions/execution.clj:
```diff
@@ -77,50 +44,104 @@
+                     400
+                     (tru "Missing primary key parameter: {0}"
+                          (pr-str (u/slugify (:name pk-field)))))
+        _ (api/check (seq row-parameters)
+                     400
+                     (tru "Implicit parameters were not provided."))
```
:+1 I don't think an update would even make sense in this case.

snoe commented on src/metabase/actions/execution.clj:
```diff
@@ -77,50 +44,104 @@
+                          :row/create)
+        arg-map (cond-> query
+                  (= implicit-action :row/create)
+                  (assoc :create-row row-parameters)
```
Will fix this.

snoe commented on src/metabase/models/action.clj:
```diff
@@ -152,12 +167,76 @@
+        actions-by-id (when-let [action-ids (not-empty (keep :action_id (vals model-action-by-model-slug)))]
+                        (m/index-by :id (select-actions :id [:in action-ids])))
+        implicit-card-ids (set (map :card_id (remove :action_id model-actions)))
+        implicit-cards (if known-models
```
yes, I'll fix the naming.

snoe commented on src/metabase/api/model_action.clj:
```diff
@@ -0,0 +1,54 @@
+  "Endpoint to delete an action"
+  [model-action-id]
+  (let [action_id (db/select-field :action_id ModelAction :id model-action-id)]
+    ;; Let cascade delete handle this
+    (db/delete! Action :id action_id)
```
Creation is only separate because QueryAction is created differently than other actions. Ideally create would be combined since the separate calls leave open some risk that the second call doesn't happen. For deleting, I'd rather keep them combined to avoid that.

metamben replied:
Do I understand correctly that this is also a temporary compromise to let us prototype this faster?

snoe replied:
Pretty much, the plan is for model_action and action to merge.

snoe commented on src/metabase/models/action.clj:
```diff
@@ -152,12 +167,76 @@
+                                        :let [{:keys [table-id]} (query/query->database-and-table-ids (:dataset_query card))]
+                                        :when table-id]
+                                    [table-id (:id card)]))
+        tables (hydrate (db/select 'Table :id [:in (keys card-id-by-table-id)]) :fields)]
```
Only if the model query doesn't eventually reference a table, there's nothing stopping me from saving such a query however so good catch.

snoe committed:
Addressing Code Reviews and Scaffold changes

snoe committed:
Add type: implicit to implicit model actions

snoe committed:
Update src/metabase/api/model_action.clj

Co-authored-by: metamben <103100869+metamben@users.noreply.github.com>

snoe committed:
Merge remote-tracking branch 'origin/master' into data-apps-model-actions

snoe committed:
Allow changing card_id with PUT to dashcard for model-actions

snoe committed:
Update tests after merge

snoe committed:
Fix implicit delete and add better test coverage

snoe committed:
Addressing review changes

Add PUT test
Add schema for dashboard-id and dashcard-id params
Fix select ordering in test

snoe commented on resources/migrations/000_migrations.yaml:
```diff
@@ -12825,6 +12825,80 @@ databaseChangeLog:
+              - column:
+                  name: action_id
+                  type: int
+                  remarks: The associated action
```
I can see how this could be better but this is tricky to change and still be able to work on the related FE branches without md5 errors. This would be going away in phase 2 as well.

metamben committed:
Merge branch 'master' into data-apps-model-actions

iethree committed:
Model Actions: Frontend (#25646)

* save actions in models

* list model actions on detail page (#25648)

* Link Dashcards to model-based actions (#25770)

* Link dashcards to model-based actions

* address review comments

* Data apps model actions implicit creator (#25807)

* Link dashcards to model-based actions

* address review comments

* Link dashcards to model-based actions

* Allow implicit actions to be used like explicit ones

* address review comments

Co-authored-by: Anton Kulyk <kuliks.anton@gmail.com>

snoe committed:
Merge remote-tracking branch 'origin/master' into data-apps-model-actions

kdoh committed:
Merge branch 'master' into data-apps-model-actions
